File: leetcode/101. Symmetric Tree.py
# Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).

# For example, this binary tree [1,2,2,3,4,4,3] is symmetric:

#     1
#    / \
#   2   2
#  / \ / \
# 3  4 4  3
# But the following [1,2,2,null,3,null,3] is not:
#     1
#    / \
#   2   2
#    \   \
#    3    3
# Note:
# Bonus points if you could solve it both recursively and iteratively.

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
import queue
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        # recursively
        # left tree is same as right tree
        # Runtime: 44 ms, faster than 68.17% of Python3 online submissions for Symmetric Tree.
        # Memory Usage: 13.2 MB, less than 5.61% of Python3 online submissions for Symmetric Tree.
        if root == None:
            return True
        # elif root.left==None and root.right==None:
        #     return True
        # elif root.left==None or root.right==None:
        #     return False
        else: 
            return self.isSymmetricDFS(root.left,root.right)
    def isSymmetricDFS(self,root1,root2):
        if root1==None and root2==None:
            return True
        elif root1==None or root2==None:
            return False
        else:
            if root1.val!=root2.val:
                return False
            else:
                return self.isSymmetricDFS(root1.left,root2.right) and self.isSymmetricDFS(root1.right,root2.left)
    def isSymmetric1(self, root: TreeNode) -> bool:
        # iteratively
        if root == None:
            return True
        q1 = queue.Queue()
        q2 = queue.Queue()
        q1.put(root.left)
        q2.put(root.right)
        while q1.qsize()!=0 and q2.qsize()!=0:
            node1 = q1.get()
            node2 = q2.get()
            if node1==None and node2==None:
                continue
            elif node1==None or node2==None:
                return False
            elif node1.val != node2.val:
                return False
            else:
                q1.put(node1.left)
                q1.put(node1.right)
                q2.put(node2.right)
                q2.put(node2.left)
        return True

File: leetcode/101. Symmetric Tree.cpp
#include <iostream>
#include <unordered_set>
#include <algorithm>
#include <vector>
#include <string>
#include <numeric>
using namespace std;


//* Definition for a binary tree node.
struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

class Solution {
public:
	bool isSymmetric(TreeNode *root) {
		if (!root) return true;
		return isSymmetric(root->left, root->right);
	}
	bool isSymmetric(TreeNode *left, TreeNode *right) {
		if (!left && !right) return true;
		if (left && !right || !left && right || left->val != right->val) return false;
		return isSymmetric(left->left, right->right) && isSymmetric(left->right, right->left);
	}

};

class Solution1 {
public:
	bool isSymmetric(TreeNode* root) {
		if (!root) return true;
		queue<TreeNode*> q1, q2;
		q1.push(root->left);
		q2.push(root->right);
		while (!q1.empty() && !q2.empty()) {
			TreeNode *node1 = q1.front(); q1.pop();
			TreeNode *node2 = q2.front(); q2.pop();
			if (!node1 && !node2) continue;
			if ((node1 && !node2) || (!node1 && node2)) return false;
			if (node1->val != node2->val) return false;
			q1.push(node1->left);
			q1.push(node1->right);
			q2.push(node2->right);
			q2.push(node2->left);
		}
		return true;
	}
};